### DISCUSSION TITLE
create a dataset of multiple heterogeneous graphs

### DISCUSSION THREAD

**Discussion by Z-Rajaei** [pyg-team/pytorch_geometric#6373] [Q&A]
Hello, I want to create a dataset containing some heterogeneous graphs.

I want to create the graphs referring this link: https://github.com/pyg-team/pytorch_geometric/blob/master/examples/hetero/load_csv.py
But, here there is just one graph.

How can I change the code to specify the graph number each node is located in?
Is there any complete example?

**Comment by wsad1** [comment]
If your graph is stored across multiple csvs, could create a batch tensor as you load each csv
```
batch_list: List[Tensor] = []
x_list: List[Tensor] = []
# load csvs for a given node type
for i, csv in enumerate(csvs):
    x, _ = load_node_csv(csv, ...)
    batch = torch.ones(x.size(-1))+i
    x_list.append(x)
    bathc_list.append(batch)
data = HeteroData()
data[node_type].x = torch.cat(x_list)
data[node_type].batch = torch.cat(batch_list)
```

**Comment by Z-Rajaei** [comment]
I have stored the graph data just as the aforementioned link, except that I have another column for graph number in the nodes data files.

**Reply by wsad1** [reply]
Ah I see then in that case you could modify `load_node_csv`, assuming you know the index of the column with graph number.
```
def load_node_csv(path, index_col, encoders=None, graph_number_col_index, **kwargs):
    df = pd.read_csv(path, index_col=index_col, **kwargs)

    mapping = {index: i for i, index in enumerate(df.index.unique())}

    x = None
    if encoders is not None:
        xs = [encoder(df[col]) for col, encoder in encoders.items()]
        x = torch.cat(xs, dim=-1)
   batch = torch.tensor(df[graph_number_col_index].value)
    return x, mapping, batch
....
...
data[node_type].batch = batch
```

**Reply by Z-Rajaei** [reply]
Hi @wsad1 
I tried to implement this code.
But I get "raise KeyError(key) from err"
in line     batch = torch.tensor(df[graph_number_col_index].value)
```
def load_node_csv(path, index_col, encoders=None, graph_number_col_index = 1, **kwargs):
    df = pd.read_csv(path, index_col=index_col, **kwargs)

    mapping = {index: i for i, index in enumerate(df.index.unique())}
    print("mapping is", mapping)
    x = None
    if encoders is not None:
        xs = [encoder(df[col]) for col, encoder in encoders.items()]
        x = torch.cat(xs, dim=-1)
    batch = torch.tensor(df[graph_number_col_index].value)
    return x, mapping, batch

```

My csv file is as follows:
```
nodeId               graphnumber            name
1                               1                         boy
2                              2                          father
3                              2                          son
4                              2                          mother
```

I tried different ways of getting the value by index with no success.
I appreciate your help.

**Reply by wsad1** [reply]
Try using `df["graphnumer"]` instead

**Comment by Z-Rajaei** [comment]
Thank you @wsad1 
Lets see my graph. 
I have graphs like the image below.

![mygraph](https://user-images.githubusercontent.com/61725728/211502489-796a43ff-116d-4e6d-9c53-e2a8fd694dcc.PNG)

I should have 6 cvs files for node data and 12 csv file for edge data? Is it right?
In this way the number of files increases. Does it affect on the performance?
Is there any better way to create the csv files?

I have another question. Should I apply specific actions if there is null values for some node attributes?

**Comment by wsad1** [comment]
I think this is a clean way to store files and its the right number of node/edge files. its hard to say if it'll affect performance, you could try reading each file in a different thread if you feel single threaded performance is slow. 

There are several ways to handle null values, just listing a few you can try out 1. you could drop rows with null values if there aren't too many 2. replace null values with mean/median of column 3. replace null value with some constant like 0.
